File: server/src/controllers/roomController.ts
import { Response } from 'express';
import { Room } from '../models';
import { AuthRequest } from '../middleware/auth';
import { validateRoomIdFormat, normalizeRoomId } from '../utils/roomIdGenerator';

// ✅ Create room
export const createRoom = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    if (!req.user) {
      res.status(401).json({ success: false, message: 'Not authenticated' });
      return;
    }

    const { name, language, isPublic } = req.body;
    console.log('🔵 Creating room with isPublic:', isPublic, 'type:', typeof isPublic);

    const room = await Room.create({
      name: name || 'Untitled Room',
      ownerId: req.user.id,
      language: language || 'javascript',
      isPublic: isPublic === true, // Explicitly convert to boolean, defaults to false
      users: [
        {
          userId: req.user.id,
          userName: req.user.username || 'Unknown',
          role: 'owner',
          joinedAt: new Date(),
        },
      ],
    });

    // Generate shareable link
    const shareableLink = `${process.env.CLIENT_URL || 'http://localhost:5173'}/rooms/${room.roomId}`;

    res.status(201).json({
      success: true,
      message: 'Room created successfully',
      data: {
        room,
        roomId: room.roomId,
        shareableLink
      },
    });
  } catch (error: any) {
    console.error('❌ Error creating room:', error);
    res.status(400).json({ success: false, message: error.message || 'Failed to create room' });
  }
};

// ✅ Get all rooms
export const getRooms = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    const { public: publicRooms, owner } = req.query;
    let query: any = {};

    if (publicRooms === 'true') query.isPublic = true;
    else if (req.user && owner === 'me') query.ownerId = req.user.id;
    else if (req.user) {
      query.$or = [
        { ownerId: req.user.id },
        { collaborators: req.user.id },
        { isPublic: true },
      ];
    } else query.isPublic = true;

    const rooms = await Room.find(query)
      .populate('ownerId', 'username email avatar')
      .populate('collaborators', 'username email avatar')
      .sort({ updatedAt: -1 });

    res.json({ success: true, data: { rooms } });
  } catch (error: any) {
    res.status(500).json({ success: false, message: error.message || 'Failed to fetch rooms' });
  }
};

// ✅ Get single room by roomId
export const getRoomById = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    const { roomId } = req.params;
    console.log('📩 Fetching room:', roomId);

    const room = await Room.findOne({ roomId })
      .populate('ownerId', 'username email avatar')
      .populate('collaborators', 'username email avatar');

    if (!room) {
      console.log('❌ Room not found for:', roomId);
      res.status(404).json({ success: false, message: 'Room not found' });
      return;
    }

    console.log('🔍 Room details:', { roomId: room.roomId, name: room.name, isPublic: room.isPublic });

    if (!room.isPublic && req.user?.id !== room.ownerId.toString()) {
      const isCollab = room.collaborators.some(
        (c: any) => c._id.toString() === req.user?.id
      );
      if (!isCollab) {
        res.status(403).json({ success: false, message: 'Access denied' });
        return;
      }
    }

    res.json({ success: true, data: { room } });
  } catch (error: any) {
    console.error('🔥 Error in getRoomById:', error);
    res.status(500).json({ success: false, message: error.message || 'Failed to fetch room' });
  }
};

// ✅ Update room
export const updateRoom = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    if (!req.user) {
      res.status(401).json({ success: false, message: 'Not authenticated' });
      return;
    }

    const { roomId } = req.params;
    const { name, language, code, isPublic } = req.body;

    const room = await Room.findOne({ roomId });
    if (!room) {
      res.status(404).json({ success: false, message: 'Room not found' });
      return;
    }

    if (room.ownerId.toString() !== req.user.id) {
      res.status(403).json({ success: false, message: 'Only room owner can update' });
      return;
    }

    if (name !== undefined) room.name = name;
    if (language !== undefined) room.language = language;
    if (code !== undefined) room.code = code;
    if (isPublic !== undefined) room.isPublic = isPublic;

    await room.save();

    const updated = await Room.findById(room._id)
      .populate('ownerId', 'username email avatar')
      .populate('collaborators', 'username email avatar');

    res.json({
      success: true,
      message: 'Room updated successfully',
      data: { room: updated },
    });
  } catch (error: any) {
    console.error('❌ Error updating room:', error);
    res.status(400).json({ success: false, message: error.message || 'Failed to update room' });
  }
};

// ✅ Add collaborator
export const addCollaborator = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    if (!req.user) {
      res.status(401).json({ success: false, message: 'Not authenticated' });
      return;
    }

    const { roomId } = req.params;
    const { userId } = req.body;

    const room = await Room.findOne({ roomId });
    if (!room) {
      res.status(404).json({ success: false, message: 'Room not found' });
      return;
    }

    if (room.ownerId.toString() !== req.user.id) {
      res.status(403).json({ success: false, message: 'Only room owner can add collaborators' });
      return;
    }

    if (room.collaborators.includes(userId as any)) {
      res.status(400).json({ success: false, message: 'User is already a collaborator' });
      return;
    }

    room.collaborators.push(userId);
    await room.save();

    const updated = await Room.findById(room._id)
      .populate('ownerId', 'username email avatar')
      .populate('collaborators', 'username email avatar');

    res.json({
      success: true,
      message: 'Collaborator added successfully',
      data: { room: updated },
    });
  } catch (error: any) {
    res.status(400).json({ success: false, message: error.message || 'Failed to add collaborator' });
  }
};

// ✅ Join room
export const joinRoom = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    if (!req.user) {
      res.status(401).json({ success: false, message: 'Not authenticated' });
      return;
    }

    const { roomId } = req.params;

    // Validate Room ID format
    const normalizedRoomId = normalizeRoomId(roomId);
    if (!validateRoomIdFormat(normalizedRoomId)) {
      res.status(400).json({
        success: false,
        message: 'Invalid room ID format. Expected format: ABC-123'
      });
      return;
    }

    const room = await Room.findOne({ roomId: normalizedRoomId });
    if (!room) {
      res.status(404).json({ success: false, message: 'Room not found' });
      return;
    }

    // Check if room is active
    if (!room.isActive) {
      res.status(403).json({ success: false, message: 'Room is no longer active' });
      return;
    }

    const isOwner = room.ownerId.toString() === req.user!.id;
    const isCollab = room.collaborators.some((id) => id.toString() === req.user!.id);

    // Note: Anyone with the Room ID can join (public or private)
    // The isPublic flag only affects dashboard visibility, not join access

    // Check if user is already in the room
    const existingUser = room.users.find(u => u.userId.toString() === req.user!.id);
    if (existingUser) {
      res.json({
        success: true,
        message: 'Already in room (rejoined successfully)',
        data: { room },
      });
      return;
    }

    // Check room capacity
    if (room.users.length >= room.maxUsers) {
      res.status(403).json({
        success: false,
        message: `Room is full (${room.users.length}/${room.maxUsers} users)`
      });
      return;
    }

    // Add user to room
    if (!isOwner && !isCollab) {
      room.collaborators.push(req.user!.id as any);
    }

    room.users.push({
      userId: req.user!.id as any,
      userName: req.user!.username || 'Unknown',
      role: isOwner ? 'owner' : 'collaborator',
      joinedAt: new Date(),
    });

    await room.save();

    const updated = await Room.findById(room._id)
      .populate('ownerId', 'username email avatar')
      .populate('collaborators', 'username email avatar');

    res.json({
      success: true,
      message: 'Joined room successfully',
      data: { room: updated },
    });
  } catch (error: any) {
    res.status(400).json({ success: false, message: error.message || 'Failed to join room' });
  }
};

// ✅ Leave room
export const leaveRoom = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    if (!req.user) {
      res.status(401).json({ success: false, message: 'Not authenticated' });
      return;
    }

    const { roomId } = req.params;
    const room = await Room.findOne({ roomId });
    if (!room) {
      res.status(404).json({ success: false, message: 'Room not found' });
      return;
    }

    if (room.ownerId.toString() === req.user.id) {
      res.status(400).json({
        success: false,
        message: 'Room owner cannot leave. Delete the room instead.',
      });
      return;
    }

    const index = room.collaborators.findIndex((id) => id.toString() === req.user!.id);
    if (index === -1) {
      res.status(400).json({ success: false, message: 'You are not a collaborator' });
      return;
    }

    room.collaborators.splice(index, 1);
    await room.save();

    res.json({ success: true, message: 'Left room successfully' });
  } catch (error: any) {
    res.status(400).json({ success: false, message: error.message || 'Failed to leave room' });
  }
};

// ✅ Delete room
export const deleteRoom = async (req: AuthRequest, res: Response): Promise<void> => {
  try {
    if (!req.user) {
      res.status(401).json({ success: false, message: 'Not authenticated' });
      return;
    }

    const { roomId } = req.params;
    const room = await Room.findOne({ roomId });
    if (!room) {
      res.status(404).json({ success: false, message: 'Room not found' });
      return;
    }

    if (room.ownerId.toString() !== req.user.id) {
      res.status(403).json({ success: false, message: 'Only room owner can delete' });
      return;
    }

    await Room.findByIdAndDelete(room._id);
    res.json({ success: true, message: 'Room deleted successfully' });
  } catch (error: any) {
    res.status(500).json({ success: false, message: error.message || 'Failed to delete room' });
  }
};

File: server-spring/src/main/java/com/codeeditor/server/controller/RoomController.java
package com.codeeditor.server.controller;

import com.codeeditor.server.dto.ApiResponse;
import com.codeeditor.server.model.Room;
import com.codeeditor.server.repository.RoomRepository;
import com.codeeditor.server.util.RoomIdGenerator;
import jakarta.servlet.http.HttpServletRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.time.Instant;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/rooms")
public class RoomController {

    private static final Logger log = LoggerFactory.getLogger(RoomController.class);
    private final RoomRepository roomRepository;

    @Value("${app.cors.allowed-origins:http://localhost:5173}")
    private String clientUrls;

    public RoomController(RoomRepository roomRepository) {
        this.roomRepository = roomRepository;
    }

    @PostMapping
    public ResponseEntity<ApiResponse> createRoom(HttpServletRequest request, @RequestBody Map<String, Object> body) {
        try {
            if (!isAuth(request)) return unauth();
            String userId = uid(request);
            String username = (String) request.getAttribute("username");

            Boolean isPublic = body.get("isPublic") instanceof Boolean b ? b : false;
            Boolean reqApproval = body.get("requireApproval") instanceof Boolean b ? b : true;

            String roomId;
            do { roomId = RoomIdGenerator.generateRoomId(); } while (roomRepository.existsByRoomId(roomId));

            Room room = new Room();
            room.setRoomId(roomId);
            room.setName(body.getOrDefault("name", "Untitled Room").toString());
            room.setOwnerId(userId);
            room.setLanguage(body.getOrDefault("language", "javascript").toString());
            room.setPublic(isPublic);
            room.setRequireApproval(isPublic ? reqApproval : false);

            Room.RoomUser ownerUser = new Room.RoomUser();
            ownerUser.setUserId(userId);
            ownerUser.setUserName(username != null ? username : "Unknown");
            ownerUser.setRole("owner");
            ownerUser.setJoinedAt(Instant.now());
            room.setUsers(new ArrayList<>(List.of(ownerUser)));
            room = roomRepository.save(room);

            String clientUrl = clientUrls.split(",")[0].trim();
            String shareableLink = clientUrl + "/rooms/" + room.getRoomId();

            return ResponseEntity.status(HttpStatus.CREATED)
                    .body(ApiResponse.success("Room created successfully",
                            Map.of("room", room, "roomId", room.getRoomId(), "shareableLink", shareableLink)));
        } catch (Exception e) {
            return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage()));
        }
    }

    @GetMapping
    public ResponseEntity<ApiResponse> getRooms(HttpServletRequest request,
            @RequestParam(value = "public", required = false) String pub,
            @RequestParam(value = "owner", required = false) String owner) {
        try {
            List<Room> rooms;
            if ("true".equals(pub)) rooms = roomRepository.findByIsPublicTrueOrderByUpdatedAtDesc();
            else if (isAuth(request) && "me".equals(owner)) rooms = roomRepository.findByOwnerIdOrderByUpdatedAtDesc(uid(request));
            else if (isAuth(request)) rooms = roomRepository.findAccessibleRooms(uid(request));
            else rooms = roomRepository.findByIsPublicTrueOrderByUpdatedAtDesc();
            return ResponseEntity.ok(ApiResponse.success(Map.of("rooms", rooms)));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error(e.getMessage()));
        }
    }

    @GetMapping("/{roomId}")
    public ResponseEntity<ApiResponse> getRoomById(HttpServletRequest request, @PathVariable String roomId) {
        try {
            var opt = roomRepository.findByRoomId(roomId);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("Room not found"));
            Room room = opt.get();
            if (!room.isPublic() && isAuth(request)) {
                String userId = uid(request);
                if (!room.getOwnerId().equals(userId) && !room.getCollaborators().contains(userId))
                    return ResponseEntity.status(403).body(ApiResponse.error("Access denied"));
            }
            return ResponseEntity.ok(ApiResponse.success(Map.of("room", room)));
        } catch (Exception e) {
            return ResponseEntity.status(500).body(ApiResponse.error(e.getMessage()));
        }
    }

    @PutMapping("/{roomId}")
    public ResponseEntity<ApiResponse> updateRoom(HttpServletRequest request, @PathVariable String roomId, @RequestBody Map<String, Object> body) {
        try {
            if (!isAuth(request)) return unauth();
            var opt = roomRepository.findByRoomId(roomId);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("Room not found"));
            Room room = opt.get();
            if (!room.getOwnerId().equals(uid(request))) return ResponseEntity.status(403).body(ApiResponse.error("Only room owner can update"));
            if (body.containsKey("name")) room.setName(body.get("name").toString());
            if (body.containsKey("language")) room.setLanguage(body.get("language").toString());
            if (body.containsKey("code")) room.setCode(body.get("code").toString());
            if (body.containsKey("isPublic")) { boolean p = (Boolean) body.get("isPublic"); room.setPublic(p); if (!p) room.setRequireApproval(false); }
            if (body.containsKey("requireApproval")) room.setRequireApproval(room.isPublic() ? (Boolean) body.get("requireApproval") : false);
            room = roomRepository.save(room);
            return ResponseEntity.ok(ApiResponse.success("Room updated successfully", Map.of("room", room)));
        } catch (Exception e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @PostMapping("/{roomId}/add-collaborator")
    public ResponseEntity<ApiResponse> addCollaborator(HttpServletRequest request, @PathVariable String roomId, @RequestBody Map<String, String> body) {
        try {
            if (!isAuth(request)) return unauth();
            var opt = roomRepository.findByRoomId(roomId);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("Room not found"));
            Room room = opt.get();
            if (!room.getOwnerId().equals(uid(request))) return ResponseEntity.status(403).body(ApiResponse.error("Only room owner can add collaborators"));
            String cId = body.get("userId");
            if (room.getCollaborators().contains(cId)) return ResponseEntity.badRequest().body(ApiResponse.error("User is already a collaborator"));
            room.getCollaborators().add(cId);
            room = roomRepository.save(room);
            return ResponseEntity.ok(ApiResponse.success("Collaborator added successfully", Map.of("room", room)));
        } catch (Exception e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @PostMapping("/{roomId}/join")
    public ResponseEntity<ApiResponse> joinRoom(HttpServletRequest request, @PathVariable String roomId) {
        try {
            if (!isAuth(request)) return unauth();
            String userId = uid(request);
            String username = (String) request.getAttribute("username");
            String nid = RoomIdGenerator.normalizeRoomId(roomId);
            if (!RoomIdGenerator.validateRoomIdFormat(nid)) return ResponseEntity.badRequest().body(ApiResponse.error("Invalid room ID format. Expected format: ABC-123"));
            var opt = roomRepository.findByRoomId(nid);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("Room not found"));
            Room room = opt.get();
            if (!room.isActive()) return ResponseEntity.status(403).body(ApiResponse.error("Room is no longer active"));
            boolean isOwner = room.getOwnerId().equals(userId);
            boolean isCollab = room.getCollaborators().contains(userId);
            boolean existing = room.getUsers().stream().anyMatch(u -> u.getUserId().equals(userId));
            if (existing) return ResponseEntity.ok(ApiResponse.success("Already in room (rejoined successfully)", Map.of("room", room)));
            if (room.getUsers().size() >= room.getMaxUsers()) return ResponseEntity.status(403).body(ApiResponse.error(String.format("Room is full (%d/%d users)", room.getUsers().size(), room.getMaxUsers())));
            if (room.isRequireApproval() && !isOwner && !isCollab) {
                return ResponseEntity.ok(ApiResponse.builder().success(false).requireApproval(true).message("Waiting for owner approval...").data(Map.of("roomId", room.getRoomId(), "roomName", room.getName(), "ownerId", room.getOwnerId())).build());
            }
            if (!isOwner && !isCollab) room.getCollaborators().add(userId);
            Room.RoomUser nu = new Room.RoomUser(); nu.setUserId(userId); nu.setUserName(username != null ? username : "Unknown"); nu.setRole(isOwner ? "owner" : "collaborator"); nu.setJoinedAt(Instant.now());
            room.getUsers().add(nu);
            room = roomRepository.save(room);
            return ResponseEntity.ok(ApiResponse.success("Joined room successfully", Map.of("room", room)));
        } catch (Exception e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @PostMapping("/{roomId}/leave")
    public ResponseEntity<ApiResponse> leaveRoom(HttpServletRequest request, @PathVariable String roomId) {
        try {
            if (!isAuth(request)) return unauth();
            String userId = uid(request);
            var opt = roomRepository.findByRoomId(roomId);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("Room not found"));
            Room room = opt.get();
            if (room.getOwnerId().equals(userId)) return ResponseEntity.badRequest().body(ApiResponse.error("Room owner cannot leave. Delete the room instead."));
            if (!room.getCollaborators().contains(userId)) return ResponseEntity.badRequest().body(ApiResponse.error("You are not a collaborator"));
            room.getCollaborators().remove(userId);
            roomRepository.save(room);
            return ResponseEntity.ok(ApiResponse.builder().success(true).message("Left room successfully").build());
        } catch (Exception e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @DeleteMapping("/{roomId}")
    public ResponseEntity<ApiResponse> deleteRoom(HttpServletRequest request, @PathVariable String roomId) {
        try {
            if (!isAuth(request)) return unauth();
            var opt = roomRepository.findByRoomId(roomId);
            if (opt.isEmpty()) return ResponseEntity.status(404).body(ApiResponse.error("Room not found"));
            Room room = opt.get();
            if (!room.getOwnerId().equals(uid(request))) return ResponseEntity.status(403).body(ApiResponse.error("Only room owner can delete"));
            roomRepository.deleteById(room.getId());
            return ResponseEntity.ok(ApiResponse.builder().success(true).message("Room deleted successfully").build());
        } catch (Exception e) { return ResponseEntity.status(500).body(ApiResponse.error(e.getMessage())); }
    }

    private boolean isAuth(HttpServletRequest r) { return Boolean.TRUE.equals(r.getAttribute("authenticated")); }
    private String uid(HttpServletRequest r) { return (String) r.getAttribute("userId"); }
    private ResponseEntity<ApiResponse> unauth() { return ResponseEntity.status(401).body(ApiResponse.error("Not authenticated")); }
}
